src/hl/selection.rs

```diff
@@ -0,0 +1,1476 @@
+        match self {
+            Self::None => {
+                h5check(H5Sselect_none(space_id))?;
+            }
+            Self::All => {
+                h5check(H5Sselect_all(space_id))?;
+            }
+            Self::Points(ref coords) => set_points_selection(space_id, coords.view())?,
+            Self::RegularHyperslab(ref hyper) => set_regular_hyperslab(space_id, hyper)?,
+            Self::ComplexHyperslab => fail!("Complex hyperslabs are not supported"),
+        };
+        Ok(())
+    }
+
+    pub unsafe fn extract_from_dataspace(space_id: hid_t) -> Result<Self> {
+        Ok(match H5Sget_select_type(space_id) {
+            H5S_sel_type::H5S_SEL_NONE => Self::None,
+            H5S_sel_type::H5S_SEL_ALL => Self::All,
+            H5S_sel_type::H5S_SEL_POINTS => Self::Points(get_points_selection(space_id)?),
+            H5S_sel_type::H5S_SEL_HYPERSLABS => get_regular_hyperslab(space_id)?
+                .map_or(Self::ComplexHyperslab, Self::RegularHyperslab),
+            sel_type => fail!("Invalid selection type: {:?}", sel_type as c_int),
+        })
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SliceOrIndex {
+    Index(isize),
+    Slice { start: isize, step: isize, end: Option<isize>, block: bool },
+    Unlimited { start: isize, step: isize, block: bool },
+}
+
+impl SliceOrIndex {
+    pub fn to_unlimited(self) -> Result<Self> {
+        Ok(match self {
+            Self::Index(_) => fail!("Cannot make index selection unlimited"),
+            Self::Slice { end: Some(_), .. } => {
+                fail!("Cannot make bounded slice unlimited")
+            }
+            Self::Slice { start, step, end: None, block } => Self::Unlimited { start, step, block },
+            Self::Unlimited { .. } => self,
+        })
+    }
+
+    pub fn to_block(self) -> Result<Self> {
+        Ok(match self {
+            Self::Index(_) => fail!("Cannot make index selection block-like"),
+            Self::Slice { start, step, end, .. } => Self::Slice { start, step, end, block: true },
+            Self::Unlimited { start, step, .. } => Self::Unlimited { start, step, block: true },
+        })
+    }
+
+    pub fn is_index(self) -> bool {
+        matches!(self, Self::Index(_))
+    }
+
+    pub fn is_slice(self) -> bool {
+        matches!(self, Self::Slice { .. })
+    }
+
+    pub fn is_unlimited(self) -> bool {
+        matches!(self, Self::Unlimited { .. })
+    }
+}
+
+#[allow(clippy::fallible_impl_from)]
+impl<T: Into<ndarray::SliceInfoElem>> From<T> for SliceOrIndex {
+    fn from(slice: T) -> Self {
+        match slice.into() {
+            ndarray::SliceInfoElem::Index(index) => Self::Index(index),
+            ndarray::SliceInfoElem::Slice { start, end, step } => {
+                Self::Slice { start, step, end, block: false }
+            }
+            ndarray::SliceInfoElem::NewAxis => panic!("ndarray NewAxis can not be mapped to hdf5"),
```

Comment:
This is a bit troublesome for conversion from the new `SliceInfo` struct. `ndarray` is OK with panicing on indexing failures, but we could try to use the `TryFrom` trait everywhere instead. Maybe something to fix at a later time?

Reply:
Yea... also:

- `impl<T, Din, Dout> From<ndarray::SliceInfo<T, Din, Dout>> for Hyperslab` 
  relies on `From<ndarray::SliceIndex>` so it would have to become a `TryFrom` as well
- Whatever uses that will have to become a `TryFrom`
- `impl From<Array1<Ix>> for Selection` also contains an `.unwrap()`

It's not nice having a core part of API potentially panicking whereas we return `Result<>` almost everywhere so we can handle it. Probably worth reviewing this entire module to see if we can avoid all panics (e.g. by using `TryFrom`).